File: insert-into-a-binary-search-tree/insert-into-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        '''
        Given a value, inserts the value into the tree.
        Returns the root of the tree the val is inserted into
        '''
        if root is None: #Tree is empty, return new tree
            return TreeNode(val)

        iter = root
        while True:
            if iter.left is None and iter.val > val: #No node, val < node.val
                iter.left = TreeNode(val)
                break
            elif iter.right is None and iter.val < val: #No node, val > node.val
                iter.right = TreeNode(val)
                break
            elif iter.left and iter.val > val: #Node exists, val < node.val
                iter = iter.left
            else: #Node exists, val > node.val
                iter = iter.right
        return root

File: insert-into-a-binary-search-tree/insert-into-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution 
{
    public TreeNode insertIntoBST(TreeNode root, int val) 
    {
        if (root == null) // No tree exists
            return new TreeNode(val);
        
        DFSInsert(root, val); // Perform insertion
        return root;
    }
    
    public void DFSInsert(TreeNode node, int val)
    {
        if (node.val > val && node.left == null)
            node.left = new TreeNode(val);
        else if (node.val > val)
            DFSInsert(node.left, val);
        else if (node.val < val && node.right == null)
            node.right = new TreeNode(val);
        else
            DFSInsert(node.right, val);
    }
}